### PR TITLE
Add functionality for path promotion to native tools bootstrapping scripts

### PR DESCRIPTION
**Pull request by jonfortescue**
Closes https://github.com/dotnet/core-eng/issues/15689.

This introduces opt-in functionality to the native tools bootstrapping scripts that does one of two things:

1) If running in an Azure Pipelines agent, promotes native tools specified in the global.json to the path.
2) If running locally, breaks the build if native tools specified in the global.json on are not present on the path.

### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/arcade/tree/main/Documentation/Validation

### CODE REVIEW

File: eng/common/init-tools-native.ps1 [hunk 1]
```diff
@@ -41,7 +41,8 @@ Param (
   [switch] $Force = $False,
   [int] $DownloadRetries = 5,
   [int] $RetryWaitTimeInSeconds = 30,
-  [string] $GlobalJsonFile
+  [string] $GlobalJsonFile,
+  [switch] $PathPromotion
```
Comment (garath) [inline eng/common/init-tools-native.ps1:45]: A `.PARAMETER` description above for this.

Comment (garath) [inline eng/common/init-tools-native.ps1:45]: Should this have a default? Any concerns about breaking existing callers?

Comment (MattGal) [reply eng/common/init-tools-native.ps1:45]: don't all switches have a default of $false?

File: eng/common/init-tools-native.ps1 [hunk 2]
```diff
@@ -77,53 +78,83 @@ try {
...
-      $LocalInstallerArguments += @{ Version = "$ToolVersion" }
-
-      if ($Verbose) {
-        $LocalInstallerArguments += @{ Verbose = $True }
-      }
-      if (Get-Variable 'Force' -ErrorAction 'SilentlyContinue') {
-        if($Force) {
-          $LocalInstallerArguments += @{ Force = $True }
+    if ($PathPromotion -eq $True) {
+      if ($env:AGENT_ID) { # check to see if we're in an Azure pipelines build
```
Comment (MattGal) [inline eng/common/init-tools-native.ps1:82]: AGENT_ID is kind of generic.  Perhaps to be more TFS-y you could check System.TeamProject / System.TeamFoundationCollectionUri

File: eng/common/init-tools-native.ps1 [hunk 3]
```diff
@@ -77,53 +78,83 @@ try {
...
+        $NativeTools.PSObject.Properties | ForEach-Object {
+          $ToolName = $_.Name
+          $ToolVersion = $_.Value
+
+          if ((Get-Command "$ToolName" -ErrorAction SilentlyContinue) -eq $null) {
+            if ($ToolVersion -eq "latest") {
+              $ToolVersion = ""
+            }
+            $ToolDirectory = (Get-ChildItem -Path "C:\arcade-tools\" -Filter "$ToolName-$ToolVersion*" | Sort-Object -Descending)[0]
```
Comment (MattGal) [inline eng/common/init-tools-native.ps1:91]: is this "latest version goes into a path named `c:\arcade-tools\$toolName-`" documented anywhere?  What does this do when c:\arcade-tools isn't defined?

File: eng/common/init-tools-native.ps1 (hunk 1 above)
Comment (garath) [reply eng/common/init-tools-native.ps1:45]: ... 🤦

**Review by garath — APPROVED** [review]

File: eng/common/init-tools-native.ps1 [hunk 4]
```diff
@@ -77,53 +78,83 @@ try {
...
+            $BinPath = Get-Content "$($ToolDirectory.FullName)\binpath.txt"
```
Comment (MattGal) [inline eng/common/init-tools-native.ps1:92]: This should check that this file exists too, we're being way too trusting of whoever makes this folder

**Review by MattGal — COMMENTED** [review]
I'd like to see more null / existence checking please

File: eng/common/init-tools-native.ps1 (hunk 3 above)
Comment (jonfortescue) [reply eng/common/init-tools-native.ps1:91]: Not yet -- I'll add that documentation as part of my Helix Machines PR.

**Review by garath — APPROVED** [review]

**Pull request merged**